Dev logic between filter's component and the calendar grid

Xamilu opened the pull request.

Xamilu committed:
feat: add custom icon button to component's folder

Xamilu committed:
feat: create the calendar section with the grid and courses cell

Xamilu committed:
feat(calendarfilters): implement the logic of the calendar navigation

Xamilu committed:
feat: create the interaction for the calendar

Create the logic between filters and dates on the grid. Manage the logic of switching weeks in a
month. Display courses correctly on the grid

IlhanYapici commented on src/components/IconButton/IconButton-types.ts:
Pas besoin d'un fichier `*-types.ts` pour juste 1 interface, tu peux la mettre dans le fichier du composant directement.

IlhanYapici commented on src/components/IconButton/IconButton.tsx:
```diff
@@ -0,0 +1,12 @@
+	const { className } = props
+
+	return <Button {...props} className={classNames('icon-button', className)} />
+}
```
De cette façon, `className` n'existe plus dans `restProps`.
```suggestion
export function IconButton(props: IIconButtonProps) {
	const { className, ...restProps } = props

	return <Button {...restProps} className={classNames('icon-button', className)} />
}
```

IlhanYapici commented on src/routes/calendar/Calendar-utils.ts:
```diff
@@ -0,0 +1,40 @@
+	const firstDayOfMonth = currentDate.startOf('month')
+	const weekNumber = Math.ceil((currentDate.date() + firstDayOfMonth.day()) / 7)
+
+	let startOfWeek = currentDate.startOf('week').add(1, 'day').subtract(1, 'day')
```
Pourquoi ajouter puis retirer un jour ?

IlhanYapici commented on src/routes/calendar/_components/calendar-courses/CalendarCourses-utils.ts:
```diff
@@ -0,0 +1,31 @@
+		padding: '8px',
+		overflow: 'hidden',
+	}
+}
```
Je pense que ça aurait été plus simple (et + précis) de positionner les cours dans une colonne si cette dernière était une grid.

IlhanYapici commented on src/routes/calendar/_components/calendar-filters/CalendarFilters-styles.less:
```diff
@@ -0,0 +1,26 @@
+	width: 100%;
+	justify-content: center;
+	gap: 5%;
+	padding: 20px 0px;
```
Évite d'utiliser des pourcentages pour tout ce qui est gap, margin, padding, etc... privilégie également les variables antd. Exemple: `--ant-margin` et `--ant-padding` (`--ant-gap` n'éxiste pas) avec les variantes xs, sm, lg, xl --> `--ant-margin-xs`.

IlhanYapici commented on src/routes/calendar/_components/calendar-filters/CalendarFilters-styles.less:
```diff
@@ -0,0 +1,26 @@
+.calendar-filter-container {
+	display: flex;
```
Pourquoi mettre `display: flex;` alors que le composant concerné par cette classe est déjà en display flex ?

IlhanYapici commented on src/routes/calendar/_components/calendar-filters/CalendarFilters.tsx:
```diff
@@ -0,0 +1,51 @@
+import { LeftOutlined, RightOutlined } from '@ant-design/icons'
```
On utilise une autre librairie d'icones `lucide-react` ([liste des icones](https://lucide.dev/icons/)).

IlhanYapici commented on src/routes/calendar/_components/calendar-filters/CalendarFilters.tsx:
Il serait peut être intéressant d'ajouter un bouton pour revenir à la date actuelle (i.e. aujourd'hui).

IlhanYapici commented on src/routes/calendar/_components/calendar-grid/CalendarGrid-types.ts:
```diff
@@ -0,0 +1,30 @@
+export interface ICalendarEvent {
+	id: string
+	title: string
+	startTime: string
+	endTime: string
+	classroom: string
+	date: string
+	color: string
+}
```
Il y a déjà des types pour les models API `src/types/api/*`. Tu peux te baser sur ceux-ci pour construire les types de tes composants, exemple:
```typescript
import { ModelAPI } from '@types/api/example'

export interface ExampleInterface extends ModelAPI {
  color: string
}
```

IlhanYapici reviewed (commented):
# Quelques remarques globales :

### Problème d'affichage

Le calendrier ne s'adapte pas à mon écran.
<img width="2020" alt="Screenshot 2025-02-15 at 16 30 03" src="https://github.com/user-attachments/assets/b3095a1a-6c85-4592-8bb3-3e04eb7f99da" />


### Rendre le composant plus "modulaire"

C'est à dire pouvoir passer un [type générique](https://www.typescriptlang.org/docs/handbook/2/generics.html) au composant et pouvoir customiser le rendu d'une cellule (`CalendarCourses`). Exemple approximatif :

```tsx
interface CalendarProps<DataType> {
  cellRender?: (cellData: DataType) => ReactNode
  data: DataType[]
}

function defaultCellRender<DataType>(data: DataType) {
  return <CalendarCourses data={data} />
}

export function Calendar<DataType>(props: CalendarProps<DataType>) {
        const { cellRender = defaultCellRender } = props
	const {
		currentDate,
		monthOptions,
		yearOptions,
		updateCurrentDate,
		handleMonthSelectChange,
		handleYearSelectChange,
	} = useCalendar()

	const weekDates = useMemo(() => getWeekDates(currentDate), [currentDate])

	return (
		<div>
			<CalendarFilters
				currentDate={currentDate}
				monthOptions={monthOptions}
				yearOptions={yearOptions}
				updateCurrentDate={updateCurrentDate}
				handleMonthSelectChange={handleMonthSelectChange}
				handleYearSelectChange={handleYearSelectChange}
			/>
			<CalendarGrid
			        cellRender={cellRender}
				events={sampleEvents}
				startHour={8}
				endHour={17}
				weekDates={weekDates}
				hourHeight={53}
			/>
		</div>
	)
}

// Exemple d'utilisation
interface Event {
	title: string
	description: string
}

function TeacherView() {
        const events: Event[] = []
	return (
		<Calendar<Event>
		        data={events}
			cellRender={(event) => (
				<Space>
					<Text>{event.title}</Text>
					<Text>{event.description}</Text>
				</Space>
			)}
		/>
	)
}
```

### Utiliser `display: grid;` pour le positionnement d'une celulle dans une colonne

Je pense qu'utiliser une grid pour positionner les cellules dans une colonnes aurait faciliter les choses (positionnement, taille d'une cellule, etc...), que ce soit pour la colonne d'heure ou pour les colonnes de cours.

### Commente ton code

Pas besoin d'écire des pavés. Quand une fonction/composant est relativement simple, commente juste la fonction/composant. Lorsque ce dernier devient plus complexe, commente également le code localement. Exemple généré par notre ami Claude :

```tsx
// Simple function example
/**
 * Calculates the total price including tax for a given amount.
 * Takes a price and returns the total with a 20% tax rate applied.
 */
function calculateTotalWithTax(price) {
    return price * 1.2;
}

// Complex function example
/**
 * Processes an array of transactions and generates a summary report.
 * Groups transactions by category, calculates totals and averages,
 * and identifies any suspicious patterns based on predefined rules.
 */
function analyzeTransactions(transactions) {
    // Group transactions by category using a map for O(1) lookups
    const categorizedTransactions = new Map();
    transactions.forEach(transaction => {
        if (!categorizedTransactions.has(transaction.category)) {
            categorizedTransactions.set(transaction.category, []);
        }
        categorizedTransactions.get(transaction.category).push(transaction);
    });

    // Calculate statistics for each category
    const categoryStats = new Map();
    categorizedTransactions.forEach((transactions, category) => {
        // Sum all transactions in this category
        const total = transactions.reduce((sum, t) => sum + t.amount, 0);
        
        // Calculate average and identify outliers using standard deviation
        const amounts = transactions.map(t => t.amount);
        const average = total / transactions.length;
        const standardDev = calculateStandardDeviation(amounts, average);
        
        // Flag suspicious transactions (beyond 2 standard deviations)
        const suspiciousTransactions = transactions.filter(t => {
            return Math.abs(t.amount - average) > standardDev * 2;
        });

        categoryStats.set(category, {
            total,
            average,
            standardDev,
            suspiciousTransactions,
            transactionCount: transactions.length
        });
    });

    // Generate final report object with all statistics and flags
    return {
        categories: Array.from(categoryStats.entries()).map(([category, stats]) => ({
            category,
            ...stats
        })),
        totalTransactions: transactions.length,
        uniqueCategories: categorizedTransactions.size,
        timestamp: new Date()
    };
}

// Helper function for standard deviation calculation
function calculateStandardDeviation(values, mean) {
    const squareDiffs = values.map(value => {
        const diff = value - mean;
        return diff * diff;
    });
    const avgSquareDiff = squareDiffs.reduce((sum, diff) => sum + diff, 0) / values.length;
    return Math.sqrt(avgSquareDiff);
}
```

### N'oublie pas les tests unitaires !

Identifie des fonctions qui mériteraient d'avoir un ou plusieurs tests unitaires.

Xamilu committed:
refactor(calendar): move calendar to components directory and make it generic

- Move Calendar to src/components
- Calendar accepts generic types
- Mandatory cellRender props
